pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,128 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+)
+
+type DfsNode struct {
+	expanded bool // true once all node neighbors are added to queue
+	parent   string
+	depth    int
+}
+
+// TODO: make more robust usuing predicates
+func searchDependenciesFromStartNode(ctx context.Context, gqlclient graphql.Client, startID string, stopID string, maxDepth int) (map[string]DfsNode, error) {
+	startNode, err := model.Node(ctx, gqlclient, startID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed getting intial node with given ID:%v", err)
+	}
+
+	_, ok := startNode.Node.(*model.NodeNodePackage)
+
+	if !ok {
+		return nil, fmt.Errorf("Not a package")
+	}
+
+	var path []string
+	queue := make([]string, 0) // the queue of nodes in bfs
+
+	nodeMap := map[string]DfsNode{}
+
+	nodeMap[startID] = DfsNode{}
+	queue = append(queue, startID)
+
+	for len(queue) > 0 {
+		now := queue[0]
+		queue = queue[1:]
+		nowNode := nodeMap[now]
+
+		if stopID == now {
+			break
+		}
+
+		if maxDepth != 0 && nowNode.depth >= maxDepth {
+			break
+		}
+
+		isDependencyNeighborResponses, err := model.Neighbors(ctx, gqlclient, now, []model.Edge{model.EdgePackageIsDependency})
+		if err != nil {
+			return nil, fmt.Errorf("failed getting package parent:%w", err)
+		}
+		for _, neighbor := range isDependencyNeighborResponses.Neighbors {
+			if isDependency, ok := neighbor.(*model.NeighborsNeighborsIsDependency); ok {
```

Comment:
would be helpful to refactor this to case on each type of node so it is easier for one to see the behavior for each node, and then look up the behavior that way... For example, having a switch statement on type(neighbor), each calling to its own function.

Reply:
Fixed (on new PR #969)